deepseg.cc

```diff
@@ -298,15 +289,26 @@ void calc_mask(calcinfo_t &info, timinginfo_t &ti) {
 			out[n] = (val & 0xE0) | (out[n] >> 3);
 		}
 	}
-	ti.maskns=timestamp();
+	info.pti->maskns = diffnanosecs(timestamp(), t0);
 
 	// denoise
 	cv::Mat tmpbuf;
 	cv::dilate(info.ofinal,tmpbuf,info.element);
 	cv::erode(tmpbuf,info.ofinal,info.element);
 
 	// scale up into full-sized mask
-	cv::resize(info.ofinal,info.mroi,cv::Size(info.raw.rows/info.ratio,info.raw.rows));
+	pthread_mutex_lock(&info.lock_mask);
+	cv::resize(info.ofinal,info.mroi,cv::Size(raw.rows/info.ratio,raw.rows));
+	pthread_mutex_unlock(&info.lock_mask);
+}
+
+void *calc_mask_thread(void *args) {
+	calcinfo_t *calcinfo = (calcinfo_t *)args;
```

Comment:
Would we do better to use a `pthread_cond_wait()/pthread_cond_signal()` pair below to ensure mask calculation only occurs at <= video frame rate? Otherwise we guarantee to consume a core (or two!) calculating flat out. Of course the `start` flag needs to remain (possibly renamed `running`) so we can terminate the thread.

Reply:
Thanks for the hint!
I implemented it. Is this the way you were thinking of?

Reply:
As we are using C++ I'd even suggest using the thread toolkit available with C++11 and newer. This even allows for implementing proper RAII paradigms which strongly avoids issues with those primitives not appearing pairwise.

Reply:
Heh - you can tell I'm too old for modern C++ :) good shout @BenBE

Reply:
Nah, not everybody needs to be fully up-to-date on everything all the time. But it sure helps to have a look around once in a while. ;-)